### ISSUE TITLE
bearing_capacity_biopolymer

### ISSUE THREAD

**Issue by mobinaizadi400451004** [mobinaizadi400451004/bearing_capacity_biopolymer_#1]
# -*- coding: utf-8 -*-
"""
Unified workflow: Gradient Boosting + PSO-ANN
Biopolymer-modified soil bearing capacity prediction
"""

import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pyswarms as ps

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# ================= CONFIG =================
TARGET = "q_ult_kpa"
dosage_col = "biopolymer_dosage_pct"
curing_col = "curing_time_days"
drop_cols = [
    "record_id", "study_id", "soil_type", "uscs_class", "biopolymer_type",
    "biopolymer_family", "mixing_method", "ucs_kpa", "q_allow_kpa",
    "bearing_capacity_index_bci",
]

# ================= LOAD =================
candidates = ["biopolymer_bearing_capacity_1000.csv",
              "bearing_capacity_biopolymer_1000.csv"]
FILE = next((c for c in candidates if os.path.exists(c)), None)
if FILE is None:
    raise FileNotFoundError(f"No dataset found among: {candidates}")

df = pd.read_csv(FILE)
df = df.rename(columns={"BearingCapacity": TARGET, "UCS": "ucs_kpa"})
print(f"Loaded {FILE}: {df.shape}")

# ================= 1) PSO-ANN =================
Xp = df.drop(columns=[TARGET, "ucs_kpa"])
yp = df[TARGET].values.reshape(-1, 1)

cat = Xp.select_dtypes(include=["object"]).columns.tolist()
num = Xp.select_dtypes(exclude=["object"]).columns.tolist()

pre = ColumnTransformer([
    ("num", StandardScaler(), num),
    ("cat", OneHotEncoder(drop="first", sparse_output=False), cat),
])
Xp_proc = pre.fit_transform(Xp)

yp_scaler = StandardScaler()
yp_scaled = yp_scaler.fit_transform(yp)

Xp_tr, Xp_te, yp_tr, yp_te = train_test_split(Xp_proc, yp_scaled, test_size=0.2, random_state=42)

input_dim = Xp_tr.shape[1]
h1, h2 = 16, 8

w1_end = input_dim * h1
b1_end = w1_end + h1
w2_end = b1_end + h1 * h2
b2_end = w2_end + h2
w3_end = b2_end + h2

def predict(Xd, w):
    W1 = w[0:w1_end].reshape(input_dim, h1)
    b1 = w[w1_end:b1_end].reshape(h1)
    W2 = w[b1_end:w2_end].reshape(h1, h2)
    b2 = w[w2_end:b2_end].reshape(h2)
    W3 = w[b2_end:w3_end].reshape(h2, 1)
    b3 = w[w3_end:w3_end + 1].reshape(1)
    return np.tanh(np.tanh(Xd @ W1 + b1) @ W2 + b2) @ W3 + b3

def mse_cost(weights):
    return np.array([np.mean((predict(Xp_tr, w) - yp_tr) ** 2) for w in weights])

dims = input_dim * h1 + h1 + h1 * h2 + h2 + h2 + 1
opt = ps.single.GlobalBestPSO(n_particles=50, dimensions=dims,
                              options={"c1": 1.494, "c2": 1.494, "w": 0.729})
best_cost, best_pos = opt.optimize(mse_cost, iters=250)

y_pred_pso = yp_scaler.inverse_transform(predict(Xp_te, best_pos))
y_te_pso = yp_scaler.inverse_transform(yp_te)

pso_r2 = r2_score(y_te_pso, y_pred_pso)
pso_mae = mean_absolute_error(y_te_pso, y_pred_pso)
pso_rmse = np.sqrt(mean_squared_error(y_te_pso, y_pred_pso))
print(f"[PSO-ANN] R2={pso_r2:.4f} | MAE={pso_mae:.2f} | RMSE={pso_rmse:.2f}")

# ================= 2) GRADIENT BOOSTING =================
drop = [c for c in drop_cols if c in df.columns]
Xg = df.drop(columns=drop + [TARGET], errors="ignore").select_dtypes(include=[np.number])
Xg = Xg.dropna(axis=1, how="all").fillna(Xg.mean(numeric_only=True))
yg = df[TARGET]
feature_cols = list(Xg.columns)

Xg_tr, Xg_te, yg_tr, yg_te = train_test_split(Xg, yg, test_size=0.2, random_state=42)

gbr = GradientBoostingRegressor(n_estimators=300, learning_rate=0.05,
                                max_depth=4, random_state=42)
gbr.fit(Xg_tr, yg_tr)
y_pred_gbr = gbr.predict(Xg_te)

gbr_r2 = r2_score(yg_te, y_pred_gbr)
gbr_mae = mean_absolute_error(yg_te, y_pred_gbr)
gbr_rmse = np.sqrt(mean_squared_error(yg_te, y_pred_gbr))
resid = yg_te - y_pred_gbr
print(f"[GBR] R2={gbr_r2:.4f} | MAE={gbr_mae:.4f} | RMSE={gbr_rmse:.4f}")

# ================= 3) RESIDUAL HISTOGRAM =================
plt.figure(figsize=(10, 6))
sns.histplot(resid, bins=30, kde=True, color="goldenrod", edgecolor="black")
plt.axvline(0, color="red", linestyle="--", lw=2)
plt.title("Residual Error Distribution - Gradient Boosting")
plt.xlabel("Residual (kPa)")
plt.ylabel("Frequency")
plt.grid(alpha=0.3)
plt.tight_layout()
plt.savefig("gbr_residual_histogram.png", dpi=300, bbox_inches="tight")
plt.show()

# ================= 4) ACTUAL VS PREDICTED =================
plt.figure(figsize=(7, 7))
plt.scatter(yg_te, y_pred_gbr, alpha=0.7, color="darkgreen", edgecolor="black")
plt.plot([yg_te.min(), yg_te.max()], [yg_te.min(), yg_te.max()], "r--", lw=2)
plt.xlabel("Actual q_ult (kPa)")
plt.ylabel("Predicted q_ult (kPa)")
plt.title("Actual vs Predicted - Gradient Boosting")
plt.grid(alpha=0.3)
plt.tight_layout()
plt.savefig("gbr_actual_vs_predicted.png", dpi=300, bbox_inches="tight")
plt.show()

# ================= 5) MODEL COMPARISON =================
models = ["PSO-ANN", "GBR"]
r2_scores = [pso_r2, gbr_r2]
mae_scores = [pso_mae, gbr_mae]

fig, ax1 = plt.subplots(figsize=(10, 6))
bars = ax1.bar(models, r2_scores, color=["#A5C9CA", "#395B64"], alpha=0.85, width=0.55)
ax1.set_xlabel("Model")
ax1.set_ylabel("R2 Score", color="tab:blue")
ax1.tick_params(axis="y", labelcolor="tab:blue")
ax1.set_ylim(0, 1.1)
for bar in bars:
    ax1.text(bar.get_x() + bar.get_width() / 2, bar.get_height() + 0.02,
             f"{bar.get_height():.3f}", ha="center", va="bottom", fontweight="bold")

ax2 = ax1.twinx()
ax2.plot(models, mae_scores, color="tab:red", marker="o", markersize=9, lw=2.5, ls="--")
ax2.set_ylabel("MAE (kPa)", color="tab:red")
ax2.tick_params(axis="y", labelcolor="tab:red")
ax2.set_ylim(0, max(mae_scores) * 1.15)
plt.title("PSO-ANN vs Gradient Boosting")
plt.tight_layout()
plt.savefig("model_comparison_summary.png", dpi=300, bbox_inches="tight")
plt.show()

# ================= 6) FEATURE IMPORTANCE =================
fi = pd.Series(gbr.feature_importances_, index=feature_cols).sort_values(ascending=False)
top_n = min(15, len(fi))

plt.figure(figsize=(10, 7))
sns.barplot(x=fi.iloc[:top_n].values, y=fi.iloc[:top_n].index, palette="viridis")
plt.title(f"Top {top_n} Feature Importances - Gradient Boosting")
plt.xlabel("Importance")
plt.ylabel("Features")
plt.tight_layout()
plt.savefig("gbr_feature_importance.png", dpi=300, bbox_inches="tight")
plt.show()
print(fi.head(15))

# ================= 7) SENSITIVITY ANALYSIS =================
base = Xg_tr.mean(numeric_only=True).to_dict()

def run_sensitivity(col, color, fname, xlabel):
    if col not in feature_cols:
        print(f"[skip] {col} not found in features")
        return
    rng = np.linspace(Xg_tr[col].min(), Xg_tr[col].max(), 60)
    rows = [dict(base, **{col: v}) for v in rng]
    sens = pd.DataFrame(rows).reindex(columns=feature_cols)
    pred = gbr.predict(sens)
    plt.figure(figsize=(10, 6))
    plt.plot(rng, pred, color=color, lw=3)
    plt.title(f"Sensitivity: {col}")
    plt.xlabel(xlabel)
    plt.ylabel("Predicted q_ult (kPa)")
    plt.grid(alpha=0.3)
    plt.tight_layout()
    plt.savefig(fname, dpi=300, bbox_inches="tight")
    plt.show()

run_sensitivity(dosage_col, "darkred", "sensitivity_biopolymer_dosage.png", "Dosage (%)")
run_sensitivity(curing_col, "navy", "sensitivity_curing_time.png", "Curing time (days)")

# ================= 8) EXPORT =================
fi.to_csv("gbr_feature_importance.csv", header=["importance"])

print("\n=== FINAL ===")
print(f"PSO-ANN: R2={pso_r2:.4f} | MAE={pso_mae:.2f} | RMSE={pso_rmse:.2f}")
print(f"GBR    : R2={gbr_r2:.4f} | MAE={gbr_mae:.4f} | RMSE={gbr_rmse:.4f}")
